packages/beacon-node/src/chain/blocks/verifyBlocksExecutionPayloads.ts

```diff
@@ -41,14 +43,102 @@ export async function verifyBlocksExecutionPayload(
+  //
+  //     Or in other words if
+  //       - block status is syncing
+  //       - and we are not in a post merge world and is parent is not optimistically safe
+  //       - and we are syncing close to the chain head i.e. clock slot
+  //       - and parent is optimistically safe
+  //
+  //      then throw error
+  //
+  //
+  //       - if we are are not yet in post merge world (preState0's isMergeTransitionComplete
+  //         is false) and
+  //       - if we haven't yet imported a post merge ancestor in forkchoice i.e.
+  //       - and we are syncing close to the clockSlot, i.e. merge Transition could be underway
+  //
+  //
+  // 2. The current slot (as per the system clock) is at least
+  //    SAFE_SLOTS_TO_IMPORT_OPTIMISTICALLY ahead of the slot of the block being
+  //    imported.
+  //    This means that the merge transition could be underway and we can't afford to import
+  //    a block which is not fully validated as it could affect liveliness of the network.
+  //
+
+  // For this segment of blocks:
+  //   We are optimistically safe with respec to this entire block segment if:
+  //    - we have transitioned to post-merge world or
+  //    - all the blocks are way behind the current slot
+  //    - or we have already imported a post-merge parent of first block of this chain in forkchoice
+  const lastBlock = blocks[blocks.length - 1];
+  let isOptimisticallySafe =
+    (isBellatrixStateType(preState0) && isMergeTransitionComplete(preState0 as BeaconStateBellatrix)) ||
+    lastBlock.message.slot + opts.safeSlotsToImportOptimistically < chain.clock.currentSlot;
+
+  if (!isOptimisticallySafe) {
+    const firstBlock = blocks[0];
+    const parentRoot = toHexString(firstBlock.message.parentRoot);
+    const parentBlock = chain.forkChoice.getBlockHex(parentRoot);
+    if (!parentBlock) {
+      throw new BlockError(firstBlock, {
+        code: BlockErrorCode.PARENT_UNKNOWN,
+        parentRoot,
+      });
+    }
+
+    // isOptimisticallySafe if we have already imported a post-merge parent of this block
+    if (parentBlock.executionStatus !== ExecutionStatus.PreMerge) {
+      isOptimisticallySafe = true;
+    }
+  }
+
   for (const block of blocks) {
     // If blocks are invalid in consensus the main promise could resolve before this loop ends.
     // In that case stop sending blocks to execution engine
     if (signal.aborted) {
       throw new ErrorAborted("verifyBlockExecutionPayloads");
     }
 
-    const {executionStatus} = await verifyBlockExecutionPayload(chain, block, preState0, opts);
+    const {executionStatus} = await verifyBlockExecutionPayload(chain, block, preState0, opts, isOptimisticallySafe);
+    // It becomes optimistically  safefor following blocks if a post-merge block is deemed fit
+    // for import. If it would not have been safe verifyBlockExecutionPayload would throw error
+    // and we would not be here.
+    isOptimisticallySafe ||= executionStatus !== ExecutionStatus.PreMerge;
```

Comment:
While the `||=` is pretty cool, can you use a regular if statement since it's more readable?